File: CS_Projects_Py_CPP_Swift_SQL/coursework_weeks/cs50x/week1/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long number = get_long("Number: ");

    if (number < 0)
    {
        printf("INVALID\n");
        return 0;
    }

    long temp = number;
    int sum = 0;
    int position = 0;
    int length = 0;
    int first_digit = 0;
    int first_two = 0;

    while (temp > 0)
    {
        int digit = temp % 10;

        if (position % 2 == 1)
        {
            int product = digit * 2;
            sum += product / 10 + product % 10;
        }
        else
        {
            sum += digit;
        }

        length++;
        if (temp < 10)
        {
            first_digit = temp;
        }
        else if (temp < 100)
        {
            first_two = temp;
        }

        temp /= 10;
        position++;
    }

    if (sum % 10 != 0)
    {
        printf("INVALID\n");
        return 0;
    }

    if (length == 15 && (first_two == 34 || first_two == 37))
    {
        printf("AMEX\n");
    }
    else if (length == 16 && (first_two >= 51 && first_two <= 55))
    {
        printf("MASTERCARD\n");
    }
    else if ((length == 13 || length == 16) && first_digit == 4)
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

File: CS_Projects_Py_CPP_Swift_SQL/coursework_weeks/cs50x/week6/sentimental-credit/credit.py
# TODO
from cs50 import get_string
import re

def main():
    number = get_string("Number: ")
    
    if is_valid(number):
        print_brand(number)
    else:
        print("INVALID")

def is_valid(number):
    # Luhn's Algorithm
    digits = [int(d) for d in number]
    checksum = 0
    
    # Process digits starting from second to last, moving left
    for i in range(len(digits) - 2, -1, -2):
        doubled = digits[i] * 2
        checksum += doubled // 10 + doubled % 10
        
    # Add the remaining digits
    for i in range(len(digits) - 1, -1, -2):
        checksum += digits[i]
        
    return checksum % 10 == 0

def print_brand(number):
    length = len(number)
    
    if length == 15 and re.match(r"^3[47]", number):
        print("AMEX")
    elif length == 16 and re.match(r"^5[1-5]", number):
        print("MASTERCARD")
    elif (length == 13 or length == 16) and re.match(r"^4", number):
        print("VISA")
    else:
        print("INVALID")

if __name__ == "__main__":
    main()
